packages/pkl.experimental.structuredRead/structuredRead.pkl

```diff
@@ -0,0 +1,261 @@
+//===----------------------------------------------------------------------===//
+// Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//===----------------------------------------------------------------------===//
+module pkl.experimental.structuredRead.structuredRead
+import "pkl:reflect"
+import "@deepToTyped/deepToTyped.pkl"
+
+/// Takes a given class or module and attempts to fill the properties using values
+/// read using the `read()` function and the set `inputScheme` (default is `prop:`)
+///
+/// Read values are coerced into the types specified in the class or module, making
+/// it easy to get external properties (passed using the `-p` flag) or environment
+/// variables (if `inputScheme` is `env:`) into a structured and typed format.
+///
+/// When embeded in another application, this Pkl module can be used to provide a
+/// basic ability to override loaded configuration using CLI flags. The application
+/// will need to parse the passed flags, then provide them to Pkl via a custom
+/// [Resource Reader](https://pkl-lang.org/main/current/language-reference/index.html#extending-resource-readers),
+/// or by passing them in as external properties.
+function fill(target: Class|Module|Dynamic) =
+  let (targetClass =
+    if (!(target is Class))
+      target.getClass()
+    else
+      target
+  )
+    let (mappedInputs =
+      if (targetClass == Dynamic)
+        applyToDynamic(List(), target)
+      else
+        applyToClass(targetClass, List()))
+      let (failures = findFailures(mappedInputs))
+      if (failures.length > 0)
+        throwFailures(targetClass, failures)
+      else
+        if (targetClass != Dynamic)
+          deepToTyped.apply(targetClass, mappedInputs)
+        else
+          mappedInputs
+
+/// Sets the scheme used with `read()`. By default is `prop:` so values
+/// are read from external properties passed using the `-p` flag.
+///
+/// If set to `env:` this would result in values being read from
+/// environment variables.
+hidden inputScheme: String(matches(Regex(#"^[\d\w-]+:$"#)))
```

Comment:
Personal preference. I don't like it when the `:` is implied because not every program does that, and you end wondering if you need to provide it or not. Making it explicit just makes it clear who's responsible for providing the `:`. One less thing to consider if you're debugging some that using this module.